pexpect/serial_spawn.py

```diff
@@ -0,0 +1,93 @@
+import os
+import sys
+import time
+import threading
+from pexpect.fdpexpect import SpawnBase
+from pexpect.exceptions import EOF, TIMEOUT, ExceptionPexpect
+
+PY3 = (sys.version_info[0] >= 3)
+
+
+class SerialSpawn(SpawnBase):
+    '''This is a serial interface class interface for Pexpect. Use this class to
+    control an existing python serial connection. '''
+    def __init__(self, ser, args=None, **kwargs):
+        if not ser.is_open:
+            raise ExceptionPexpect('serial port is not ready')
+        super().__init__(None, **kwargs)
+        self.ser = ser
+        self.name = f'<serial port {ser.port}>'
+        self.closed = False
+        self.interacting = threading.Event()
+
+    def close(self):
+        if not self.ser.is_open:
+            return
+        self.flush()
+        self.ser.close()
+        self.closed = True
+
+    def isalive(self):
+        return self.ser.is_open
+
+    def send(self, s):
+        s = self._coerce_send_string(s)
+        self._log(s, 'send')
+        b = self._encoder.encode(s, final=False)
+        return self.ser.write(b)
+    
+    def read_nonblocking(self, size=1, timeout=-1):
+        if not self.isalive():
+            raise EOF('End Of File (EOF).')
+        if timeout and timeout < 0:
+            timeout = self.timeout
+        self.ser.timeout = timeout
+        s = self.ser.read(n)
```

Comment:
[pyserial timeout](https://pyserial.readthedocs.io/en/latest/pyserial_api.html)

Possible values for the parameter timeout which controls the behavior of [read()](https://pyserial.readthedocs.io/en/latest/pyserial_api.html#serial.Serial.read):

timeout = None: wait forever / until requested number of bytes are received
timeout = 0: non-blocking mode, return immediately in any case, returning zero or more, up to the requested number of bytes
timeout = x: set timeout to x seconds (float allowed) returns immediately when the requested number of bytes are available, otherwise wait until the timeout expires and return all bytes that were received until then.


Not sure what will happen if timeout=-1.
please look at the example usage of read() in code: https://github.com/pyserial/pyserial/blob/31fa4807d73ed4eb9891a88a15817b439c4eea2d/serial/threaded/__init__.py#L198

data = self.serial.read(self.serial.in_waiting or 1)

Reply:
Avoid the undefined behaviour. Do not wait in case of a negative value for timeout.